modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/coordination/rdbms/DatabaseSlotAgent.java

```diff
@@ -25,244 +29,603 @@
+            }
+        }
+        return allSlotsForQueue;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void reassignSlot(Slot slotToBeReassigned) throws AndesException {
+
+        for (int attemptCount = 1; attemptCount <= MAX_STORE_FAILURE_TOLERANCE_COUNT; attemptCount++) {
+
+            waitUntilStoresBecomeAvailable();
+            try {
+                andesContextStore.deleteSlotAssignment(slotToBeReassigned.getStartMessageId(),
+                        slotToBeReassigned.getEndMessageId());
+                break;
+            } catch (AndesStoreUnavailableException e) {
+                handleFailure(attemptCount, "delete slot assignment with start message id: "
+                                            + slotToBeReassigned.getStartMessageId() + " and end message id: "
+                                            + slotToBeReassigned.getEndMessageId(), e);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void deleteOverlappedSlots(String nodeId) throws AndesException {
+        //Not necessary in RDBMS mode, because RDBMS mode uses single table to store information
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void updateOverlappedSlots(String nodeId, String queueName, TreeSet<Slot> overlappedSlots)
+            throws AndesException {
+
+        //We don't need to check whether the store is available here since it is handled in the lower level method
+        for (Slot slot : overlappedSlots) {
+            this.setSlotState(slot.getStartMessageId(), slot.getEndMessageId(), SlotState.OVERLAPPED);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<String> getAllQueues() throws AndesException {
+
+        Set<String> allQueues = null;
+        for (int attemptCount = 1; attemptCount <= MAX_STORE_FAILURE_TOLERANCE_COUNT; attemptCount++) {
+
+            waitUntilStoresBecomeAvailable();
+            try {
+                allQueues = andesContextStore.getAllQueues();
+                break;
+            } catch (AndesStoreUnavailableException e) {
+                handleFailure(attemptCount, "retrieve all queue", e);
+            }
+        }
+        return allQueues;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void clearSlotStorage() throws AndesException {
+
+        for (int attemptCount = 1; attemptCount <= MAX_STORE_FAILURE_TOLERANCE_COUNT; attemptCount++) {
+
+            waitUntilStoresBecomeAvailable();
+            try {
+                andesContextStore.clearSlotStorage();
+                break;
+            } catch (AndesStoreUnavailableException e) {
+                handleFailure(attemptCount, "clear slot storage", e);
+            }
+        }
+    }
+
+    /**
+     * Method to block the thread until the stores become available.
+     */
+    private void waitUntilStoresBecomeAvailable() throws AndesException {
```

Comment:
Could we pass an argument which indicates while during which task outage happened. ( than printing a generic log message)